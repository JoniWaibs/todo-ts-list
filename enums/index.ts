export enum ItemState {
  DONE = 'done',
  TODO = 'todo',
  WIP = 'wip',
};