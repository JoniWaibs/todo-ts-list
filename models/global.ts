import { RenderResult } from '@testing-library/react';

export interface Global {
  appRenderWrapper(
    ui: JSX.Element,
    appProps?: any
  ): RenderResult,
}