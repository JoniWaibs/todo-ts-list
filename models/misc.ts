export interface StylableComponent {
  customClass?: string,
}