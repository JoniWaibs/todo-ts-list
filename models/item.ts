import { ItemState } from "@/enums/index";

export interface Item {
  id: string,
  name: string,
  state: ItemState,
}

export interface ListItem {
  items: Item[];
}