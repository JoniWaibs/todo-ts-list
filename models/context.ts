import { Dispatch, SetStateAction } from 'react';
import { Item } from './item';

export interface AppContextModel {
  items: Item[],
  setItems: Dispatch<SetStateAction<Item[]>>,
  doCompleteItem: (id: string) => void,
  deleteItem: (id: string) => void,
}

export interface AppProviderProps {
  children: JSX.Element | JSX.Element[]
}