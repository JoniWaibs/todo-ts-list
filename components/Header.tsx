import React, { useContext, useState } from 'react'
import { v4 as uuidv4 } from 'uuid';
import { mapContent } from '@/content/index';
import { AppContext } from '@/context/App';
import { ItemState } from '@/enums/index';
import styled from 'styled-components';

const { TODO } = ItemState;

const HeaderContainer = styled.header`
  padding-bottom: 45px;
`

const Title = styled.h1`
  text-align: center;
  padding: 25px 0;
`;

const ContainerInput = styled.div`
  display: flex;
  width: 100%
`;

const StyledInput = styled.input`
  width: 100%;
  padding:15px 10px;
  border-radius: 10px 0 0 10px;
  border: transparent;
  font-size: 16px;
  &:input:focus {
    outline: none;
    box-shadow: 0px 0px 2px red;
  }
`;

const SubmitButton = styled.button`
  cursor: pointer;
  border-radius: 0 10px 10px 0;
  border: transparent;
  width: 20%;
  font-weight: 700;
  font-size: 16px;
  background-color: #479aff;
  color: #d9fafc;
`;

export const Header: React.FunctionComponent = () => {
  const [input, setInput] = useState<string>('');
  const {items, setItems} = useContext(AppContext);

  const handleSubmit = () => {
    if(!input) return;
    setItems([...items, { id: uuidv4(), name: input, state: TODO }])
  }

  const { addYourTask, done } = mapContent();

  return (
    <HeaderContainer>
      <Title>TODO list</Title>
      <ContainerInput>
        <StyledInput
          type="text"
          role="input"
          aria-label={addYourTask}
          placeholder={addYourTask}
          onChange={(e: {target: HTMLInputElement}) => setInput(e.target.value)}
          value={input}
        />
        <SubmitButton role="add-task-button" onClick={handleSubmit}>{done}</SubmitButton>
      </ContainerInput>
    </HeaderContainer>
  )
}
