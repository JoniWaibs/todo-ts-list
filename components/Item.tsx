import { useContext } from 'react';
import styled from 'styled-components';
import { AppContext } from '@/context/App';
import { ItemState } from '@/enums/index';
import { Item as ItemProps } from '@/models/item';
import { CheckIcon } from '@/svg/CheckIcon';
import { TrashIcon } from '@/svg/TrashIcon';

const { DONE } = ItemState;

const RowItem = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  width: 100%;
  gap: 5px;
  background-color: #ffffff;
  border-radius: 5px;
  padding: 10px
`;

const DeleteButton = styled.button`
  border-radius: 5px;
  border:transparent;
  background-color: #f4675d
`;

const DoneButton = styled.button<{state: string}>`
  border-radius: 5px;
  border:transparent;
  background-color: ${props => props.state === DONE ? '#bfbfbf' : '#47ab67'};
`;

const ItemInfo = styled.div `
  display: flex;
  align-items: center;
  width: 100%
`;

const TaskName = styled.p<{state: string}>`
  padding: 0 15px;
  font-size: 18px;
  text-decoration: ${props => props.state === DONE ? "line-through" : 'none'};
`;

const StateLabel = styled.span<{state?: string}>`
  background-color: ${props => props.state === DONE ? '#a8f5c1' : '#ffffff'};
  color: green;
  font-weight: bold;
  padding: 2px 10px;
  text-transform: uppercase;
  font-size: 12px;
  border-radius: 15px;
`;

export const Item = ({name, state, id}: ItemProps): JSX.Element => {
  const { doCompleteItem, deleteItem } = useContext(AppContext)

  const handleCompleteItem = (id: string) => {
    if(!id) return null;
    doCompleteItem(id)
  };

  const handleDeleteItem = (id: string) => {
    if(!id) return null;
    deleteItem(id)
  };

  return (
    <RowItem>
      <DoneButton role="complete-button" onClick={() => handleCompleteItem(id)} state={state}>
        <CheckIcon customClass={'#ffff'}/>
      </DoneButton>
      <ItemInfo>
        <TaskName state={state}>{name}</TaskName>
        <StateLabel state={state} >{state === DONE ? state : null}</StateLabel>
      </ItemInfo>
      <DeleteButton role="delete-button" onClick={() => handleDeleteItem(id)}>
        <TrashIcon customClass={'#ffff'}/>
      </DeleteButton>
    </RowItem>
  );
}
