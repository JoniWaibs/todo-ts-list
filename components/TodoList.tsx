import React, { useContext } from 'react'
import { AppContext } from '@/context/App';
import { Item as ItemModel } from '@/models/item';
import { Item } from './Item';
import styled from 'styled-components';

const ContainerList = styled.div`
  width: 100%;
  display: flex;
  flex-direction: column;
  gap: 15px;
`;

export const TodoList: React.FunctionComponent = () => {
  const { items } = useContext(AppContext);

  return (
    <ContainerList>
      { items && items.map(({name, state, id}: ItemModel) => (<Item name={name} state={state} id={id} key={id}/>))}
    </ContainerList>
  )
}
