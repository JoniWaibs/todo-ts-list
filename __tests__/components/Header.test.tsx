import React, { useState as useStateMock } from 'react';
import { fireEvent, cleanup } from '@testing-library/react';
import { Item as ItemProps } from '@/models/item';
import { ItemState } from '@/enums/index';
import { Header } from "@/components/Header";
import { Global } from '@/models/global';

const { TODO, DONE } = ItemState;

declare const global: Global;

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
}));

describe('Header Component', () => {
  let items: ItemProps[] = [];
  let setItems: () => void;
  let doCompleteItem: () => void;
  let deleteItem: () => void;
  let setState: () => void;

  beforeEach(() => {
    setItems = jest.fn();
    doCompleteItem = jest.fn();
    deleteItem = jest.fn();
    setState = jest.fn();

    items = [{
      name: 'some little task',
      state: TODO,
      id: 'id-task-test',
    },
    {
      name: 'play football',
      state:  DONE,
      id: 'id-task-test',
    }];

    // Mock useState behavior
    (useStateMock as jest.Mock).mockImplementation(init => [init, setState])
  });

  afterEach(() => {
    cleanup();
  });

  test('Should input a new task to do', () => {
    const { container, getByLabelText } = global.appRenderWrapper(
      <Header />,
      {
        items,
        setItems,
        doCompleteItem,
        deleteItem,
      }
    );

    // Change input value
    const inputRangeElement = getByLabelText('Add your task');
    fireEvent.change(inputRangeElement, { target: { value: 'Task to do it tomorrow' } });

    expect(container).toMatchSnapshot();
    expect(setState).toHaveBeenCalledWith('Task to do it tomorrow')
  });

  test('Should set a new task to do in state manager', () => {
    // Mock input value
    (useStateMock as jest.Mock).mockReturnValueOnce(['task added', {}])

    const { container, getByRole } = global.appRenderWrapper(
      <Header />,
      {
        items,
        setItems,
        doCompleteItem,
        deleteItem,
      }
    );

    expect(container).toMatchSnapshot();

    const addButton = getByRole('add-task-button');
    fireEvent.click(addButton)
    expect(setItems).toHaveBeenCalled()
  });

  test('Should not set a new task to do when text input is empty', () => {
    const { container, getByLabelText, getByText } = global.appRenderWrapper(
      <Header />,
      {
        items,
        setItems,
        doCompleteItem,
        deleteItem,
      }
    );

    // Change input value to emulate useState behavior
    const inputRangeElement = getByLabelText('Add your task');
    fireEvent.change(inputRangeElement, { target: { value: '' } });

    expect(container).toMatchSnapshot();

    const addYourTaskButton = getByText('Done')
    fireEvent.click(addYourTaskButton);

    expect(setState).not.toHaveBeenCalledWith('Some text')
  });
});
