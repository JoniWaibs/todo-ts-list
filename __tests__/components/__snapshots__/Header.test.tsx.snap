// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component Should input a new task to do 1`] = `
<div>
  <div>
    <h1>
      TODO list
    </h1>
    <input
      aria-label="Add your task"
      placeholder="Add your task"
      role="input"
      type="text"
      value=""
    />
    <button
      role="add-task-button"
    >
      Done
    </button>
  </div>
</div>
`;

exports[`Header Component Should not set a new task to do when text input is empty 1`] = `
<div>
  <div>
    <h1>
      TODO list
    </h1>
    <input
      aria-label="Add your task"
      placeholder="Add your task"
      role="input"
      type="text"
      value=""
    />
    <button
      role="add-task-button"
    >
      Done
    </button>
  </div>
</div>
`;

exports[`Header Component Should set a new task to do in state manager 1`] = `
<div>
  <div>
    <h1>
      TODO list
    </h1>
    <input
      aria-label="Add your task"
      placeholder="Add your task"
      role="input"
      type="text"
      value="task added"
    />
    <button
      role="add-task-button"
    >
      Done
    </button>
  </div>
</div>
`;
